cloudify_aws/lambda_serverless/resources/function.py

```diff
@@ -79,16 +88,44 @@ def invoke(self, params):
         '''
             Invokes an AWS Lambda Function.
         '''
-        params = params or dict()
-        params.update(dict(FunctionName=self.resource_id))
-        self.logger.debug('Invoking %s with parameters: %s'
-                          % (self.type_name, params))
-        res = self.client.invoke(**params)
+        invoke_params = dict()
+        invoke_params.update(params)
+        invoke_params.update(dict(FunctionName=self.resource_id))
+        with self._encode_payload(invoke_params.get('Payload')) as \
+                encoded_payload:
+            if encoded_payload is not None:
+                invoke_params['Payload'] = encoded_payload
+            self.logger.debug('Invoking %s with parameters: %s'
+                              % (self.type_name, invoke_params))
+            res = self.client.invoke(**invoke_params)
         if res and res.get('Payload'):
-            res['Payload'] = res['Payload'].read()
+            res['Payload'] = self._decode_payload(res['Payload'])
         self.logger.debug('Response: %s' % res)
         return res
 
+    @contextmanager
+    def _encode_payload(self, payload):
+        if isinstance(payload, str):
+            with file(payload, 'r') as payload_file:
+                yield payload_file
+        elif isinstance(payload, dict):
+            yield json.dumps(payload).encode(self.resource_encoding)
+        else:
+            yield payload
+
+    def _decode_payload(self, payload_stream):
+        payload = payload_stream.read().decode(self.resource_encoding)
+        try:
+            payload = json.loads(payload)
+            if payload.get('body'):
+                try:
+                    payload['body'] = json.loads(payload['body'])
+                except ValueError as e:
+                    self.logger.warn('Cannot decode body: %s' % e)
+        except (ValueError, UnicodeDecodeError) as e:
+            self.logger.warn('Cannot decode Payload: %s' % e)
```

Comment:
Do we really care here about just rasing warning  ?

Reply:
not really, I will remove the warn

Reply:
done